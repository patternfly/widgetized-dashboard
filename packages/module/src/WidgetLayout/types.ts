import { Layout } from 'react-grid-layout';

export const widgetIdSeparator = '#';

export type Variants = 'sm' | 'md' | 'lg' | 'xl';

export type LayoutWithTitle = Layout & { title: string };

export type TemplateConfig = {
  [k in Variants]: LayoutWithTitle[];
};

export type PartialTemplateConfig = Partial<TemplateConfig>;

// Extended type the UI tracks
export type ExtendedLayoutItem = LayoutWithTitle & {
  widgetType: string;
  config?: WidgetConfiguration;
  locked?: boolean;
};

// Extended type the UI tracks
export type ExtendedTemplateConfig = {
  [k in Variants]: ExtendedLayoutItem[];
};

// Extended type the UI tracks
export type PartialExtendedTemplateConfig = Partial<ExtendedTemplateConfig>;

export interface WidgetDefaults {
  w: number;
  h: number;
  maxH: number;
  minH: number;
}

export interface WidgetHeaderLink {
  title?: string;
  href?: string;
}

export interface WidgetConfiguration {
  icon?: React.ReactNode;
  headerLink?: WidgetHeaderLink;
  title?: string;
}

/**
 * Widget definition with rendering function
 */
export interface WidgetDefinition {
  /** Unique widget type identifier */
  widgetType: string;
  /** Default dimensions for the widget */
  defaults: WidgetDefaults;
  /** Widget configuration (title, icon, header link) */
  config?: WidgetConfiguration;
  /** Function that renders the widget content */
  renderWidget: (widgetId: string) => React.ReactNode;
}

/**
 * Widget mapping keyed by widget type
 */
export interface WidgetMapping {
  [widgetType: string]: Omit<WidgetDefinition, 'widgetType'>;
}

/**
 * Notification type for user feedback
 */
export interface Notification {
  variant: 'success' | 'danger' | 'warning' | 'info' | 'default';
  title: string;
  description?: string;
}

/**
 * Analytics tracking function (optional)
 */
export type AnalyticsTracker = (event: string, data?: Record<string, unknown>) => void;

