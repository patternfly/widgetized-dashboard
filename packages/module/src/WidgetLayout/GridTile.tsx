import {
  Button,
  Card,
  CardBody,
  CardHeader,
  CardTitle,
  Divider,
  Dropdown,
  DropdownItem,
  DropdownList,
  Flex,
  FlexItem,
  HelperText,
  HelperTextItem,
  Icon,
  MenuToggle,
  MenuToggleElement,
  Skeleton,
  Tooltip,
} from '@patternfly/react-core';
import { CompressIcon, EllipsisVIcon, ExpandIcon, GripVerticalIcon, LockIcon, MinusCircleIcon, UnlockIcon } from '@patternfly/react-icons';
import React, { useMemo, useState } from 'react';
import clsx from 'clsx';
import { Link } from 'react-router-dom';
import { Layout } from 'react-grid-layout';
import { ExtendedLayoutItem, WidgetConfiguration, AnalyticsTracker } from './types';

export type SetWidgetAttribute = <T extends string | number | boolean>(id: string, attributeName: keyof ExtendedLayoutItem, value: T) => void;

export type GridTileProps = React.PropsWithChildren<{
  widgetType: string;
  icon?: React.ReactNode;
  setIsDragging: (isDragging: boolean) => void;
  isDragging: boolean;
  setWidgetAttribute: SetWidgetAttribute;
  widgetConfig: Layout & {
    colWidth: number;
    locked?: boolean;
    config?: WidgetConfiguration;
  };
  removeWidget: (id: string) => void;
  children: React.ReactNode;
  isLoaded?: boolean;
  analytics?: AnalyticsTracker;
}>;

const GridTile = ({
  widgetType,
  isDragging,
  setIsDragging,
  setWidgetAttribute,
  widgetConfig,
  removeWidget,
  children,
  isLoaded = true,
  analytics,
}: GridTileProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const { headerLink } = widgetConfig.config || {};
  const hasHeader = headerLink && headerLink.href && headerLink.title;

  const [linkHref, linkTarget] = useMemo(() => {
    if (!headerLink?.href) {
      return [];
    }

    try {
      const url = new URL(headerLink.href);
      return url.origin === window.location.origin ? [url.pathname, undefined] : [headerLink.href, '_blank'];
    } catch {
      return [headerLink.href];
    }
  }, [headerLink?.href]);

  const dropdownItems = useMemo(() => {
    const isMaximized = widgetConfig.h === widgetConfig.maxH;
    const isMinimized = widgetConfig.h === widgetConfig.minH;
    return (
      <>
        <DropdownItem
          ouiaId={widgetConfig.static ? 'unlock-widget' : 'lock-widget'}
          onClick={() => {
            setIsOpen(false);
            setWidgetAttribute(widgetConfig.i, 'static', !widgetConfig.static);
          }}
          icon={widgetConfig.static ? <UnlockIcon /> : <LockIcon />}
        >
          {widgetConfig.static ? 'Unlock location and size' : 'Lock location and size'}
        </DropdownItem>
        <DropdownItem
          ouiaId="autosize-widget"
          isDisabled={isMaximized || widgetConfig.static}
          onClick={() => {
            setWidgetAttribute(widgetConfig.i, 'h', widgetConfig.maxH ?? widgetConfig.h);
            setIsOpen(false);
          }}
          icon={<ExpandIcon />}
        >
          Autosize height to content
        </DropdownItem>
        <DropdownItem
          ouiaId="minimize-widget"
          onClick={() => {
            setWidgetAttribute(widgetConfig.i, 'h', widgetConfig.minH ?? widgetConfig.h);
            setIsOpen(false);
          }}
          isDisabled={isMinimized || widgetConfig.static}
          icon={<CompressIcon />}
        >
          Minimize height
        </DropdownItem>
        <DropdownItem
          ouiaId="remove-widget"
          onClick={() => {
            removeWidget(widgetConfig.i);
            analytics?.('widget-layout.widget-remove', { widgetType });
          }}
          icon={
            <Icon className="pf-v6-u-pb-2xl" status={widgetConfig.static ? undefined : 'danger'}>
              <MinusCircleIcon />
            </Icon>
          }
          isDisabled={widgetConfig.static}
        >
          Remove
          <HelperText>
            <HelperTextItem className="pf-v6-u-text-wrap">
              {"All 'removed' widgets can be added back by clicking the 'Add widgets' button."}
            </HelperTextItem>
          </HelperText>
        </DropdownItem>
      </>
    );
  }, [widgetConfig.h, widgetConfig.maxH, widgetConfig.minH, widgetConfig.static, widgetConfig.i, setWidgetAttribute, removeWidget, analytics, widgetType]);

  const headerActions = (
    <>
      <Tooltip content={<p>Actions</p>}>
        <Dropdown
          ouiaId={`${widgetType}-widget`}
          popperProps={{
            appendTo: document.body,
            maxWidth: '300px',
            position: 'right',
          }}
          toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
            <MenuToggle
              ref={toggleRef}
              isExpanded={isOpen}
              onClick={() => setIsOpen((prev) => !prev)}
              variant="plain"
              aria-label="widget actions menu toggle"
            >
              <EllipsisVIcon aria-hidden="true" />
            </MenuToggle>
          )}
          isOpen={isOpen}
          onOpenChange={(isOpen: boolean) => setIsOpen(isOpen)}
        >
          <DropdownList>{dropdownItems}</DropdownList>
        </Dropdown>
      </Tooltip>
      <Tooltip aria-label="Move widget" content={<p>{widgetConfig.static ? 'Widget locked' : 'Move'}</p>}>
        <Icon
          onMouseDown={() => {
            setIsDragging(true);
            analytics?.('widget-layout.widget-move', { widgetType });
          }}
          onMouseUp={() => setIsDragging(false)}
          className={clsx('drag-handle', {
            dragging: isDragging,
          })}
        >
          <GripVerticalIcon style={{ fill: 'var(--pf-t--global--icon--color--subtle)' }} />
        </Icon>
      </Tooltip>
    </>
  );

  return (
    <Card
      ouiaId={`${widgetType}-widget`}
      className={clsx('grid-tile', {
        static: widgetConfig.static,
      })}
    >
      <CardHeader className="pf-v6-u-pr-lg" actions={{ actions: headerActions }}>
        <Flex>
          <Flex className="pf-v6-u-flex-direction-row pf-v6-u-flex-nowrap">
            {widgetConfig?.config?.icon && (
              <div className="pf-v6-u-align-self-flex-start widg-c-icon--header pf-v6-u-mr-sm">
                {isLoaded ? widgetConfig.config.icon : <Skeleton shape="circle" width="25px" height="25px" />}
              </div>
            )}
            <Flex className="pf-v6-u-flex-direction-row widg-card-header-text">
              {isLoaded ? (
                <CardTitle
                  style={{
                    userSelect: isDragging ? 'none' : 'auto',
                  }}
                  className="pf-v6-u-align-self-flex-start"
                >
                  {widgetConfig?.config?.title || widgetType}
                </CardTitle>
              ) : (
                <Skeleton width="50%" />
              )}
              {hasHeader && isLoaded && (
                <FlexItem>
                  <Button
                    className="pf-v6-u-font-weight-bold pf-v6-u-font-size-xs pf-v6-u-p-0"
                    variant="link"
                    component={(props) => <Link {...props} to={linkHref} target={linkTarget} />}
                  >
                    {headerLink.title}
                  </Button>
                </FlexItem>
              )}
            </Flex>
          </Flex>
        </Flex>
      </CardHeader>
      <Divider />
      <CardBody className="pf-v6-u-p-0">{children}</CardBody>
    </Card>
  );
};

export default GridTile;

