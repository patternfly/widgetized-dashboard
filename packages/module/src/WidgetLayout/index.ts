export { default as WidgetLayout } from './WidgetLayout';
export { default as GridLayout } from './GridLayout';
export { default as GridTile } from './GridTile';
export { default as WidgetDrawer } from './WidgetDrawer';

export * from './types';
export {
  columns,
  breakpoints,
  droppingElemId,
  getWidgetIdentifier,
  mapWidgetDefaults,
  getGridDimensions,
  isWidgetType,
  extendLayout,
  mapTemplateConfigToExtendedTemplateConfig,
} from './utils';

export type { WidgetLayoutProps } from './WidgetLayout';
export type { GridLayoutProps } from './GridLayout';
export type { GridTileProps, SetWidgetAttribute } from './GridTile';
export type { WidgetDrawerProps } from './WidgetDrawer';

