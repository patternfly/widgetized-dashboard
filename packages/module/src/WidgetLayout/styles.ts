/**
 * Styles for WidgetLayout components using PatternFly tokens
 * This replaces the SCSS files with CSS-in-JS approach
 */

export const gridLayoutStyles = `
  .react-grid-item .react-resizable-handle-nw,
  .react-grid-item .react-resizable-handle-sw,
  .react-grid-item .react-resizable-handle-se {
    display: none;
  }

  .react-grid-item .react-resizable-handle-nw,
  .react-grid-item .react-resizable-handle-se {
    cursor: nwse-resize;
  }

  .react-grid-item .react-resizable-handle-ne,
  .react-grid-item .react-resizable-handle-sw {
    cursor: nesw-resize;
  }

  .react-grid-item:hover:not(.static) .react-resizable-handle-nw,
  .react-grid-item:hover:not(.static) .react-resizable-handle-sw,
  .react-grid-item:hover:not(.static) .react-resizable-handle-se,
  .react-grid-item:active:not(.static) .react-resizable-handle-nw,
  .react-grid-item:active:not(.static) .react-resizable-handle-sw,
  .react-grid-item:active:not(.static) .react-resizable-handle-se {
    display: inherit;
  }

  .react-grid-item.react-grid-placeholder {
    background-color: var(--pf-t--color--gray--60);
    border-radius: 12px;
  }

  .react-grid-item .react-resizable-handle::after {
    display: none;
  }

  .react-grid-item .react-resizable-handle img {
    padding: 3px;
  }

  #widget-layout-container {
    width: 100%;
    min-height: 200px;
  }

  .grid-tile::before {
    z-index: 1;
  }

  .grid-tile {
    background: var(--pf-t--global--background--color--100);
    position: relative;
    height: 100%;
    overflow: hidden;
  }

  .grid-tile .drag-handle {
    cursor: grab;
  }

  .grid-tile .drag-handle.dragging {
    cursor: grabbing;
  }

  .grid-tile.static .drag-handle,
  .grid-tile.static .drag-handle.dragging {
    cursor: not-allowed;
  }

  .grid-tile .pf-v6-c-card__header {
    padding: var(--pf-t--global--spacer--md);
    padding-bottom: var(--pf-t--global--spacer--sm);
    background: var(--pf-t--global--background--color--200);
  }

  .grid-tile .widg-c-icon--header .service-icon {
    height: 28px;
    width: 28px;
  }

  .grid-tile .widg-c-icon--header .pf-v6-svg {
    color: var(--pf-t--color--blue--50);
    height: 22px;
    width: 22px;
    margin-bottom: var(--pf-t--global--spacer--sm);
  }

  .grid-tile .widg-card-header-text {
    gap: 0;
    line-height: 0;
    padding-top: 2px;
  }

  .grid-tile .pf-v6-c-card__header .pf-v6-c-menu-toggle {
    padding-left: 0;
    padding-right: 0;
  }

  .grid-tile .pf-v6-c-card__actions {
    padding-left: var(--pf-t--global--spacer--xs);
  }

  .grid-tile .pf-v6-c-card__body {
    background: var(--pf-t--global--background--color--100);
  }

  .widg-c-drawer__header {
    height: 100%;
  }

  .widg-c-drawer__drag-handle {
    cursor: grab;
  }

  .widg-l-gallery {
    --pf-v6-l-gallery--m-gutter--GridGap: 8px;
  }
`;

// Inject styles into document head
if (typeof document !== 'undefined') {
  const styleId = 'widget-layout-styles';
  if (!document.getElementById(styleId)) {
    const style = document.createElement('style');
    style.id = styleId;
    style.textContent = gridLayoutStyles;
    document.head.appendChild(style);
  }
}

