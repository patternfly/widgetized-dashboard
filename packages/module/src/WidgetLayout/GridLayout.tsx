import 'react-grid-layout/css/styles.css';
import './styles';
import ReactGridLayout, { Layout, ReactGridLayoutProps } from 'react-grid-layout';
import GridTile, { SetWidgetAttribute } from './GridTile';
import { useEffect, useMemo, useRef, useState } from 'react';
import { isWidgetType } from './utils';
import React from 'react';
import {
  ExtendedLayoutItem,
  Variants,
  WidgetMapping,
  ExtendedTemplateConfig,
  AnalyticsTracker,
} from './types';
import { Button, EmptyState, EmptyStateActions, EmptyStateBody, EmptyStateVariant, PageSection } from '@patternfly/react-core';
import { ExternalLinkAltIcon, GripVerticalIcon, PlusCircleIcon } from '@patternfly/react-icons';
import { columns, breakpoints, droppingElemId, getWidgetIdentifier, extendLayout, getGridDimensions } from './utils';

export const defaultBreakpoints = breakpoints;

// SVG resize handle as inline data URI
const resizeHandleSvg = 'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTYiIGhlaWdodD0iMTYiIHZpZXdCb3g9IjAgMCAxNiAxNiIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZD0iTTE2IDEuMTQyODZMMTQuODU3MSAwTDAgMTQuODU3MVYxNkgxLjE0Mjg2TDE2IDEuMTQyODZaIiBmaWxsPSIjRDJEMkQyIi8+Cjwvc3ZnPgo=';

const getResizeHandle = (resizeHandleAxis: string, ref: React.Ref<HTMLDivElement>) => (
    <div ref={ref} className={`react-resizable-handle react-resizable-handle-${resizeHandleAxis}`}>
      <img src={resizeHandleSvg} alt="Resize handle" />
    </div>
  );

export interface GridLayoutProps {
  /** Widget mapping definition */
  widgetMapping: WidgetMapping;
  /** Current template configuration */
  template: ExtendedTemplateConfig;
  /** Callback when template changes */
  onTemplateChange?: (template: ExtendedTemplateConfig) => void;
  /** Whether the layout is locked (disables drag, drop, resize) */
  isLayoutLocked?: boolean;
  /** Custom empty state component */
  emptyStateComponent?: React.ReactNode;
  /** Documentation link for empty state */
  documentationLink?: string;
  /** Analytics tracker (optional) */
  analytics?: AnalyticsTracker;
  /** Whether to show the empty state when no widgets exist */
  showEmptyState?: boolean;
  /** Callback when drawer should be expanded */
  onDrawerExpandChange?: (expanded: boolean) => void;
  /** Currently active widgets (for tracking) */
  onActiveWidgetsChange?: (widgetTypes: string[]) => void;
}

const LayoutEmptyState = ({
  onDrawerExpandChange,
  documentationLink,
}: {
  onDrawerExpandChange?: (expanded: boolean) => void;
  documentationLink?: string;
}) => {
  useEffect(() => {
    onDrawerExpandChange?.(true);
  }, [onDrawerExpandChange]);

  return (
    <PageSection hasBodyWrapper={false} className="empty-layout pf-v6-u-p-0">
      <EmptyState headingLevel="h2" icon={PlusCircleIcon} titleText="No dashboard content" variant={EmptyStateVariant.lg} className="pf-v6-u-p-sm">
        <EmptyStateBody>
          You don't have any widgets on your dashboard. To populate your dashboard, drag <GripVerticalIcon /> items from the widget drawer to this
          dashboard.
        </EmptyStateBody>
        {documentationLink && (
          <EmptyStateActions>
            <Button variant="link" icon={<ExternalLinkAltIcon />} iconPosition="end" component="a" href={documentationLink} target="_blank">
              Learn more about widget dashboard
            </Button>
          </EmptyStateActions>
        )}
      </EmptyState>
    </PageSection>
  );
};

const GridLayout = ({
  widgetMapping,
  template,
  onTemplateChange,
  isLayoutLocked = false,
  emptyStateComponent,
  documentationLink,
  analytics,
  showEmptyState = true,
  onDrawerExpandChange,
  onActiveWidgetsChange,
}: GridLayoutProps) => {
  const [isDragging, setIsDragging] = useState(false);
  const [isInitialRender, setIsInitialRender] = useState(true);
  const [layoutVariant, setLayoutVariant] = useState<Variants>('xl');
  const [layoutWidth, setLayoutWidth] = useState<number>(1200);
  const layoutRef = useRef<HTMLDivElement>(null);

  const [currentDropInItem, setCurrentDropInItem] = useState<string | undefined>();
  const [internalTemplate, setInternalTemplate] = useState<ExtendedTemplateConfig>(template);

  // Sync external template changes to internal state
  useEffect(() => {
    setInternalTemplate(template);
  }, [template]);

  const droppingItemTemplate: ReactGridLayoutProps['droppingItem'] = useMemo(() => {
    if (currentDropInItem && isWidgetType(widgetMapping, currentDropInItem)) {
      const widget = widgetMapping[currentDropInItem];
      if (!widget) {return undefined;}
      return {
        ...widget.defaults,
        i: droppingElemId,
        widgetType: currentDropInItem,
        title: 'New title',
        config: widget.config,
      };
    }
    return undefined;
  }, [currentDropInItem, widgetMapping]);

  const setWidgetAttribute: SetWidgetAttribute = (id, attributeName, value) => {
    const newTemplate = Object.entries(internalTemplate).reduce(
      (acc, [size, layout]) => ({
        ...acc,
        [size]: layout.map((widget) => (widget.i === id ? { ...widget, [attributeName]: value } : widget)),
      }),
      {} as ExtendedTemplateConfig
    );
    setInternalTemplate(newTemplate);
    onTemplateChange?.(newTemplate);
  };

  const removeWidget = (id: string) => {
    const newTemplate = Object.entries(internalTemplate).reduce(
      (acc, [size, layout]) => ({
        ...acc,
        [size]: layout.filter((widget) => widget.i !== id),
      }),
      {} as ExtendedTemplateConfig
    );
    setInternalTemplate(newTemplate);
    onTemplateChange?.(newTemplate);
  };

  const onDrop: ReactGridLayoutProps['onDrop'] = (_layout: ExtendedLayoutItem[], layoutItem: ExtendedLayoutItem, event: DragEvent) => {
    const data = event.dataTransfer?.getData('text') || '';
    if (isWidgetType(widgetMapping, data)) {
      setCurrentDropInItem(undefined);
      const widget = widgetMapping[data];
      if (!widget) {return;}
      const newTemplate = Object.entries(internalTemplate).reduce((acc, [size, layout]) => {
        const newWidget = {
          ...layoutItem,
          ...widget.defaults,
          // make sure the configuration is valid for all layout sizes
          w: size === layoutVariant ? layoutItem.w : Math.min(widget.defaults.w, columns[size as Variants]),
          x: size === layoutVariant ? layoutItem.x : Math.min(layoutItem.x, columns[size as Variants]),
          widgetType: data,
          i: getWidgetIdentifier(data),
          title: 'New title',
          config: widget.config,
        };
        return {
          ...acc,
          [size]: layout.reduce<ExtendedLayoutItem[]>(
            (acc, curr) => {
              if (curr.x + curr.w > newWidget.x && curr.y + curr.h <= newWidget.y) {
                acc.push(curr);
              } else {
                // push the current items down on the Y axis if they are supposed to be below the new widget
                acc.push({ ...curr, y: curr.y + curr.h });
              }
              return acc;
            },
            [newWidget]
          ),
        };
      }, {} as ExtendedTemplateConfig);
      
      setInternalTemplate(newTemplate);
      onTemplateChange?.(newTemplate);
      analytics?.('widget-layout.widget-add', { data });
    }
    event.preventDefault();
  };

  const onLayoutChange = (currentLayout: Layout[]) => {
    if (isInitialRender) {
      setIsInitialRender(false);
      const activeWidgets = activeLayout.map((item) => item.widgetType);
      onActiveWidgetsChange?.(activeWidgets);
      return;
    }
    if (isLayoutLocked || currentDropInItem) {
      return;
    }

    const newTemplate = extendLayout({ ...internalTemplate, [layoutVariant]: currentLayout });
    const activeWidgets = activeLayout.map((item) => item.widgetType);
    onActiveWidgetsChange?.(activeWidgets);
    
    setInternalTemplate(newTemplate);
    onTemplateChange?.(newTemplate);
  };

  useEffect(() => {
    const currentWidth = layoutRef.current?.getBoundingClientRect().width ?? 1200;
    const variant: Variants = getGridDimensions(currentWidth);
    setLayoutVariant(variant);
    setLayoutWidth(currentWidth);
    
    const observer = new ResizeObserver((entries) => {
      if (!entries[0]) {return;}

      const currentWidth = entries[0].contentRect.width;
      const variant: Variants = getGridDimensions(currentWidth);
      setLayoutVariant(variant);
      setLayoutWidth(currentWidth);
    });
    
    if (layoutRef.current) {
      observer.observe(layoutRef.current);
    }

    return () => {
      observer.disconnect();
    };
  }, []);

  const activeLayout = internalTemplate[layoutVariant] || [];
  
  return (
    <div id="widget-layout-container" style={{ position: 'relative' }} ref={layoutRef}>
      {activeLayout.length === 0 && !currentDropInItem && showEmptyState && (
        emptyStateComponent || <LayoutEmptyState onDrawerExpandChange={onDrawerExpandChange} documentationLink={documentationLink} />
      )}
      <ReactGridLayout
        key={'grid-' + layoutVariant}
        draggableHandle=".drag-handle"
        layout={internalTemplate[layoutVariant]}
        cols={columns[layoutVariant]}
        rowHeight={56}
        width={layoutWidth}
        isDraggable={!isLayoutLocked}
        isResizable={!isLayoutLocked}
        resizeHandle={getResizeHandle as unknown as ReactGridLayoutProps['resizeHandle']}
        resizeHandles={['sw', 'nw', 'se', 'ne']}
        droppingItem={droppingItemTemplate}
        isDroppable={!isLayoutLocked}
        onDrop={onDrop}
        onDragStart={() => setCurrentDropInItem(undefined)}
        useCSSTransforms
        verticalCompact
        onLayoutChange={onLayoutChange}
      >
        {activeLayout
          .map(({ widgetType, title, ...rest }, index) => {
            const widget = widgetMapping[widgetType];
            if (!widget) {
              return null;
            }
            const config = widgetMapping[widgetType]?.config;
            return (
              <div key={rest.i} data-grid={rest} tabIndex={index} className={`widget-columns-${rest.w} widget-rows-${rest.h}`}>
                <GridTile
                  isDragging={isDragging}
                  setIsDragging={setIsDragging}
                  widgetType={widgetType}
                  widgetConfig={{ ...rest, colWidth: layoutWidth / columns[layoutVariant], config }}
                  setWidgetAttribute={setWidgetAttribute}
                  removeWidget={removeWidget}
                  analytics={analytics}
                >
                  {widget.renderWidget(rest.i)}
                </GridTile>
              </div>
            );
          })
          .filter((layoutItem) => layoutItem !== null)}
      </ReactGridLayout>
    </div>
  );
};

export default GridLayout;

