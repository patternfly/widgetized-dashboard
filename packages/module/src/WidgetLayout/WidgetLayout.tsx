import React, { useState } from 'react';
import GridLayout from './GridLayout';
import WidgetDrawer from './WidgetDrawer';
import { ExtendedTemplateConfig, WidgetMapping, AnalyticsTracker } from './types';

export interface WidgetLayoutProps {
  /** Widget mapping definition */
  widgetMapping: WidgetMapping;
  /** Initial template configuration */
  initialTemplate: ExtendedTemplateConfig;
  /** Callback when template changes */
  onTemplateChange?: (template: ExtendedTemplateConfig) => void;
  /** Whether the layout is locked (disables drag, drop, resize) */
  isLayoutLocked?: boolean;
  /** Custom empty state component */
  emptyStateComponent?: React.ReactNode;
  /** Documentation link for empty state */
  documentationLink?: string;
  /** Analytics tracker (optional) */
  analytics?: AnalyticsTracker;
  /** Whether to show the widget drawer */
  showDrawer?: boolean;
  /** Whether to show the empty state when no widgets exist */
  showEmptyState?: boolean;
  /** Custom instruction text for the drawer */
  drawerInstructionText?: string;
  /** Whether the drawer is initially open */
  initialDrawerOpen?: boolean;
};

const WidgetLayout = ({
  widgetMapping,
  initialTemplate,
  onTemplateChange,
  isLayoutLocked = false,
  emptyStateComponent,
  documentationLink,
  analytics,
  showDrawer = true,
  showEmptyState = true,
  drawerInstructionText,
  initialDrawerOpen = false,
}: WidgetLayoutProps) => {
  const [template, setTemplate] = useState<ExtendedTemplateConfig>(initialTemplate);
  const [drawerOpen, setDrawerOpen] = useState(initialDrawerOpen);
  const [currentlyUsedWidgets, setCurrentlyUsedWidgets] = useState<string[]>([]);

  const handleTemplateChange = (newTemplate: ExtendedTemplateConfig) => {
    setTemplate(newTemplate);
    onTemplateChange?.(newTemplate);
  };

  const handleDrawerExpandChange = (expanded: boolean) => {
    setDrawerOpen(expanded);
  };

  const handleActiveWidgetsChange = (widgetTypes: string[]) => {
    setCurrentlyUsedWidgets(widgetTypes);
  };

  const gridLayout = (
    <GridLayout
      widgetMapping={widgetMapping}
      template={template}
      onTemplateChange={handleTemplateChange}
      isLayoutLocked={isLayoutLocked}
      emptyStateComponent={emptyStateComponent}
      documentationLink={documentationLink}
      analytics={analytics}
      showEmptyState={showEmptyState}
      onDrawerExpandChange={handleDrawerExpandChange}
      onActiveWidgetsChange={handleActiveWidgetsChange}
    />
  );

  if (!showDrawer) {
    return gridLayout;
  }

  return (
    <WidgetDrawer
      widgetMapping={widgetMapping}
      currentlyUsedWidgets={currentlyUsedWidgets}
      isOpen={drawerOpen}
      onOpenChange={setDrawerOpen}
      instructionText={drawerInstructionText}
    >
      {gridLayout}
    </WidgetDrawer>
  );
};

export default WidgetLayout;

