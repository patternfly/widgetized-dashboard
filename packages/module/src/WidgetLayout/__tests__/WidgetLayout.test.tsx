import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { WidgetLayout } from '../WidgetLayout';
import { WidgetMapping, ExtendedTemplateConfig } from '../types';
import { CubeIcon } from '@patternfly/react-icons';
import { BrowserRouter } from 'react-router-dom';

const mockWidgetMapping: WidgetMapping = {
  'test-widget': {
    defaults: { w: 2, h: 3, maxH: 6, minH: 2 },
    config: {
      title: 'Test Widget',
      icon: <CubeIcon />
    },
    renderWidget: (id) => <div data-testid={`widget-${id}`}>Test Widget Content</div>
  }
};

const mockTemplate: ExtendedTemplateConfig = {
  xl: [
    { i: 'test-widget#1', x: 0, y: 0, w: 2, h: 3, widgetType: 'test-widget', title: 'Test Widget' }
  ],
  lg: [
    { i: 'test-widget#1', x: 0, y: 0, w: 2, h: 3, widgetType: 'test-widget', title: 'Test Widget' }
  ],
  md: [
    { i: 'test-widget#1', x: 0, y: 0, w: 2, h: 3, widgetType: 'test-widget', title: 'Test Widget' }
  ],
  sm: [
    { i: 'test-widget#1', x: 0, y: 0, w: 1, h: 3, widgetType: 'test-widget', title: 'Test Widget' }
  ]
};

const emptyTemplate: ExtendedTemplateConfig = {
  xl: [],
  lg: [],
  md: [],
  sm: []
};

describe('WidgetLayout', () => {
  it('renders without crashing', () => {
    render(
      <BrowserRouter>
        <WidgetLayout
          widgetMapping={mockWidgetMapping}
          initialTemplate={mockTemplate}
        />
      </BrowserRouter>
    );
    expect(screen.getByTestId('widget-test-widget#1')).toBeInTheDocument();
  });

  it('shows empty state when no widgets are present', () => {
    render(
      <BrowserRouter>
        <WidgetLayout
          widgetMapping={mockWidgetMapping}
          initialTemplate={emptyTemplate}
        />
      </BrowserRouter>
    );
    expect(screen.getByText(/No dashboard content/i)).toBeInTheDocument();
  });

  it('hides empty state when showEmptyState is false', () => {
    render(
      <BrowserRouter>
        <WidgetLayout
          widgetMapping={mockWidgetMapping}
          initialTemplate={emptyTemplate}
          showEmptyState={false}
        />
      </BrowserRouter>
    );
    expect(screen.queryByText(/No dashboard content/i)).not.toBeInTheDocument();
  });

  it('hides drawer when showDrawer is false', () => {
    render(
      <BrowserRouter>
        <WidgetLayout
          widgetMapping={mockWidgetMapping}
          initialTemplate={emptyTemplate}
          showDrawer={false}
          showEmptyState={false}
        />
      </BrowserRouter>
    );
    // Drawer instructions should not be present
    expect(screen.queryByText(/Add new and previously removed widgets/i)).not.toBeInTheDocument();
  });

  it('renders widget with custom title', () => {
    render(
      <BrowserRouter>
        <WidgetLayout
          widgetMapping={mockWidgetMapping}
          initialTemplate={mockTemplate}
        />
      </BrowserRouter>
    );
    expect(screen.getByText('Test Widget')).toBeInTheDocument();
  });

  it('calls onTemplateChange when template changes', () => {
    const handleChange = jest.fn();
    render(
      <BrowserRouter>
        <WidgetLayout
          widgetMapping={mockWidgetMapping}
          initialTemplate={mockTemplate}
          onTemplateChange={handleChange}
        />
      </BrowserRouter>
    );
    // Note: Testing actual drag-and-drop interactions would require more complex testing setup
    // This test just verifies the callback prop is accepted
    expect(handleChange).not.toHaveBeenCalled();
  });
});

