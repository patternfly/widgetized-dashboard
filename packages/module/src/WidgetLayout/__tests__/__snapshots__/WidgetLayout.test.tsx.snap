// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetLayout renders without crashing 1`] = `
<div>
  <div
    data-testid="widget-test-widget#1"
  >
    Test Widget Content
  </div>
</div>
`;

